### PR TITLE
gateway: spec parity

### PR DESCRIPTION
**Pull request by Arqu**
Tackles https://github.com/n0-computer/iroh/issues/277

Closes: 
- https://github.com/n0-computer/iroh/issues/230
- https://github.com/n0-computer/iroh/issues/103
- https://github.com/n0-computer/iroh/issues/197
- https://github.com/n0-computer/iroh/issues/98
- https://github.com/n0-computer/iroh/issues/89

Various other QoL fixes, most of the `ETag` support (missing dir etags), `HEAD` request handling (missing a small bit to be done), ensures headers, params etc are in line with spec.

Remainder of work will continue to be tracked on the tracker issue.

### CODE REVIEW

File: iroh-gateway/src/handlers.rs [hunk 1]
```diff
@@ -156,6 +156,9 @@ pub async fn get_handler<T: ContentLoader + std::marker::Unpin>(
         .map_err(|e| error(StatusCode::BAD_REQUEST, &e, &state))?;
     let resolved_cid = resolved_path.root();
 
+    let return_if_head = handle_only_if_cached(&request_headers, &state, &resolved_cid).await?;
```
Comment (dignifiedquire) [inline iroh-gateway/src/handlers.rs]: this is never used?

File: iroh-resolver/src/resolver.rs [hunk 2]
```diff
@@ -62,15 +62,25 @@ impl Path {
         &self.tail
     }
 
+    pub fn is_dir(&self) -> bool {
+        self.tail.len() > 0 && self.tail.last().unwrap().eq("")
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs]: That's not guranteed to be true, that's at best and indication that this is a dir

File: iroh-gateway/src/headers.rs [hunk 3]
```diff
@@ -10,16 +10,34 @@ pub fn add_user_headers(headers: &mut HeaderMap, user_headers: HeaderMap) {
...
+    let mut content_type = String::new();
+    if let Some(ct) = guess.first() {
+        content_type = ct.to_string();
+    } else {
         // todo(arqu): deeper content type checking
-        // todo(arqu): if mime type starts with text/html; strip encoding to let browser detect
-        headers.insert(
-            CONTENT_TYPE,
-            HeaderValue::from_str(content_type.as_ref()).unwrap(),
+        let classifier = mime_classifier::MimeClassifier::new();
```
Comment (dignifiedquire) [inline iroh-gateway/src/headers.rs]: could you store & reuse this classifier?

File: iroh-gateway/src/headers.rs [hunk 4]
```diff
@@ -10,16 +10,34 @@ pub fn add_user_headers(headers: &mut HeaderMap, user_headers: HeaderMap) {
...
+        let computed_type = classifier.classify(
+            context,
+            no_sniff_flag,
+            apache_bug_flag,
+            &supplied_type,
+            body_sample,
         );
     }
+
+    if content_type.starts_with("text/html") {
```
Comment (dignifiedquire) [inline iroh-gateway/src/headers.rs:26]: what case is this?

File: iroh-gateway/src/client.rs [hunk 5]
```diff
@@ -98,8 +108,15 @@ impl<T: ContentLoader + std::marker::Unpin> Client<T> {
...
                 .map_err(|e| e.to_string())?;
 
-            let stream = ReaderStream::new(reader);
-            let body = PrettyStreamBody(stream, metadata.size);
+            let mut buf_reader = tokio::io::BufReader::with_capacity(1024 * 1024, reader);
+            let body_sample = buf_reader
+                .fill_buf()
+                .await
+                .map_err(|e| e.to_string())?
+                .to_vec();
```
Comment (dignifiedquire) [inline iroh-gateway/src/client.rs:116]: Why not avoid the storing in a `Vec`, by doing the mime check  here, and store the result of it

File: iroh-resolver/src/resolver.rs (hunk 2 above)
Comment (Arqu) [reply iroh-resolver/src/resolver.rs]: Technically it's intended use is purely string/browser path based manipulation, so it is a bit misleading, especially in the resolver crate. I don't se a sensible way to have info from the resolver if its a UnixFS::Dir before we get to the actual call we want to make when just parsing the path. 
The benefit is that this is only called/used when `FileResult::Directory` is returned by the client. I could potentially rename it or just separate out this logic into the handlers not to let others lean on it?

File: iroh-gateway/src/response.rs [hunk 6]
```diff
@@ -96,37 +89,25 @@ pub fn get_response_format(
...
-                return Err("invalid format".to_string());
-            }
+    }
+
+    match ResponseFormat::try_from_headers(request_headers) {
+        Ok(format) => {
+            return Ok(format);
+        }
+        Err(_) => {
+            return Ok(ResponseFormat::Fs(String::new()));
```
[the remark this answers is not in the document]
Comment (Arqu) [reply iroh-gateway/src/response.rs]: Thought I was hallucinating. Couldn't find the code at all :D Guess you reviewed commit by commit. Clippy would have yelled at me anyways :D

File: iroh-gateway/src/headers.rs (hunk 3 above)
Comment (Arqu) [reply iroh-gateway/src/headers.rs]: It's neither large nor stateful really, we save nothing and I have to bother with a fork to add clone to it. Also having a global one would mean we have one of these cloned/instantiated per req. This way it's only when the req needs it.

File: iroh-gateway/src/client.rs (hunk 5 above)
Comment (Arqu) [reply iroh-gateway/src/client.rs:116]: and here I am shuffling 1MB around for no reason 👍

File: iroh-resolver/src/resolver.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-resolver/src/resolver.rs]: maybe call it has_trailing_slash? then it is up to the caller to give it meaning

**Pull request merged**